src/snowflake/snowpark/_internal/code_generation.py

```diff
@@ -0,0 +1,594 @@
+    """
+    Get the objects references by target func, they could be methods, modules, classes, methods, global variables
+    and its closures.
+
+    Args:
+        func: The target function to generate source code for.
+        ref_objects: dict of objects referenced by the target function, key is the name and value is the object.
+    """
+    # 1. resolve function global references
+    code_object = func.__code__
+    globals_ref = extract_func_global_refs(
+        func.__code__
+    )  # get the names of the objects which func references
+    globals = {
+        k: func.__globals__[k]
+        for k in globals_ref
+        if k in func.__globals__  # retrieve the objects by names
+        and k not in ref_objects
+    }
+
+    ref_objects.update(globals)
+
+    # 2. resolve function closure references
+    if func.__closure__ is not None:
+        closures = {
+            k: v
+            for k, v in zip(
+                code_object.co_freevars,
+                list(map(lambda x: x.cell_contents, func.__closure__)),
+            )
+        }
+        for k, v in closures.items():
+            ref_objects[k] = v
+            # if the closure item is a function and is not itself (recursive) and has not been visited
+            if isinstance(v, FunctionType) and v != func:
+                get_func_references(v, ref_objects)
+
+
+def get_class_references(
+    cls: type,
+    func: FunctionType,
+    ref_objects: Dict[str, Any],
+    classes_to_generate: list,
+    *,
+    generate_code_for_class: bool = True,
+) -> None:
+    """
+    To get the referenced objects of a class defined in the same module.
+    A class could have methods, subclasses referencing other objects.
+
+    Args:
+        cls: The class to be analyzed to find references.
+        func: The target function to generate source code for.
+        ref_objects: dict of objects referenced by the target function, key is the name and value is the object.
+        classes_to_generate: list of classes which are defined in the same module as the target function.
+            Code generation is required for these classes.
+        generate_code_for_class: Whether the source code for the class shall be generated.
+    """
+    func_module_name = func.__module__
+
+    if generate_code_for_class:
+        # order matters to classes_to_generate, when constructing source code,
+        # referenced classes need to be defined first.
+        classes_to_generate.insert(0, cls)
+
+    inferred_classes: List[Tuple[Any, bool]] = []
+
+    for base_class in cls.__bases__:
+        base_class_name = base_class.__qualname__
+        top_level_class = base_class_name.split(".")[0]
+        if base_class_name == "object":
+            continue
+        elif base_class.__module__ == func_module_name:
+            # if base class is from the same module, we need to parse the class as well as generate code for the class
+            # False in the tuple means this is not a nested class
```

Comment:
move this comment (explanation for the bool value to the line before inferred_classes definition)?

Reply:
updated, added more comments to make it clear